plugins/change-password-cyberpanel/ChangePasswordCyberPanel.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+class ChangePasswordCyberPanel implements \RainLoop\Providers\ChangePassword\ChangePasswordInterface
+{
+	/**
+	 * @var string
+	 */
+	private $mHost = '127.0.0.1';
+
+	/**
+	 * @var string
+	 */
+	private $mUser = '';
+
+	/**
+	 * @var string
+	 */
+	private $mPass = '';
+
+	/**
+	 * @var \MailSo\Log\Logger
+	 */
+	private $oLogger = null;
+
+	/**
+	 * @param string $mHost
+	 *
+	 * @return \ChangePasswordCyberPanel
+	 */
+	public function SetmHost($mHost)
+	{
+		$this->mHost = $mHost;
+		return $this;
+	}
+
+	/**
+	 * @param string $mUser
+	 *
+	 * @return \ChangePasswordCyberPanel
+	 */
+	public function SetmUser($mUser)
+	{
+		$this->mUser = $mUser;
+		return $this;
+	}
+
+	/**
+	 * @param string $mPass
+	 *
+	 * @return \ChangePasswordCyberPanel
+	 */
+	public function SetmPass($mPass)
+	{
+		$this->mPass = $mPass;
+		return $this;
+	}
+
+	/**
+	 * @param \MailSo\Log\Logger $oLogger
+	 *
+	 * @return \ChangePasswordCyberPanel
+	 */
+	public function SetLogger($oLogger)
+	{
+		if ($oLogger instanceof \MailSo\Log\Logger)
+		{
+			$this->oLogger = $oLogger;
+		}
+
+		return $this;
+	}
+
+	/**
+	 * @param \RainLoop\Account $oAccount
+	 *
+	 * @return bool
+	 */
+	public function PasswordChangePossibility($oAccount)
+	{
+		return $oAccount && $oAccount->Email();
+	}
+
+	/**
+	 * @param \RainLoop\Account $oAccount
+	 * @param string $sPrevPassword
+	 * @param string $sNewPassword
+	 *
+	 * @return bool
+	 */
+	public function ChangePassword(\RainLoop\Account $oAccount, $sPrevPassword, $sNewPassword)
+	{
+		if ($this->oLogger)
+		{
+			$this->oLogger->Write('Try to change password for '.$oAccount->Email());
+		}
+
+		$bResult = false;
+		$db = mysqli_connect($this->mHost, $this->mUser, $this->mPass, 'cyberpanel');
+
+		try
+		{
+			$sEmail = $oAccount->Email();
+			$sEmailUser = \MailSo\Base\Utils::GetAccountNameFromEmail($sEmail);
+			$sEmailDomain = \MailSo\Base\Utils::GetDomainFromEmail($sEmail);
+
+			$password_check_query = "SELECT * FROM e_users WHERE emailOwner_id = '$sEmailDomain' AND email = '$sEmail'";

```

Comment:
I think here you need to escape special characters in the values.

Reply:
Thanks for the review. Looking closely, I don't immediately see an escape character pass in the custom SQL password plug-in which I drew heavily upon in drafting this. Do you have a recommendation or preference for the method you would like this to escape special characters? I could simply use the mysqli_real_escape_string() for pre-processing the in-line variables. Is that what you are after or something more case specific?

Reply:
I added mysqli_real_escape_string processing to all relevant variables that could potentially be used in the SQL statement.

Reply:
Is there anything else you need me or CyberPanel's @usmannasir to do before you'll approve this pull request?